[#11] 주문 조회 & 주문 수락/거절

handsone-u opened the pull request:
https://github.com/Project-Baedal/store/issues/13 store 도메인에서 필요로 하는 기능을 구현했습니다.
- 주문 조회
- 주문 수락/거절

---

### 주문 수락
1. 임시 저장된 주문을 조회
2. 주문 엔티티 저장
3. 주문 상태 변경
4. 라이더 도메인에 주문 수락 전달

handsone-u committed:
[#9] correct: 인증된 사용자의 권한 유무 제거

handsone-u committed:
[#11] chore: 실행 환경 변수 추가

handsone-u committed:
[#11] feat: 주문 조회 기능

handsone-u committed:
[#13] feat: 주문 상태 변경 기능

handsone-u committed:
[#13] add: 주문 상태 변경 시 비즈니스 로직

- TODO로 표시함.

JinTaekLim reviewed (commented):
주문 환불, 등록시 배달 대기열에 등록하는 부분은 kafka 를 활용한 메세지 큐를 활용하는게 좋아보입니다!

RestAPI 로 구현하게 되면, 다른 도메인에서 주문 순서 보장이 안될뿐더러 해당 API 에서는 다른 도메인의 정보를 반환 받을 필요는 없을 것같아서요!

handsone-u committed:
[#13] add: 라이더 대기열 등록 어댑터

- kafka의 rider.addRiderQueue 토픽에서 orderId를 key로 사용하여 rider 도메인에서 처리하도록 위임

handsone-u commented:
@JinTaekLim 
kafka 활용해서 rider 도메인에 위임하도록 했습니다. [703b8d7](https://github.com/Project-Baedal/order/pull/12/commits/703b8d7a3f770053c1064c897ae145cd60101d33)

---
근데 key를 orderId로 해도 괜찮은지 싶네요. 순서 보장을 위해 다른 key를 사용해야 할까요?

JinTaekLim commented:
> @JinTaekLim kafka 활용해서 rider 도메인에 위임하도록 했습니다. [703b8d7](https://github.com/Project-Baedal/order/pull/12/commits/703b8d7a3f770053c1064c897ae145cd60101d33)
> 
> 근데 key를 orderId로 해도 괜찮은지 싶네요. 순서 보장을 위해 다른 key를 사용해야 할까요?

storeId 를 활용해 동일한 매장에 대한 순서 보장을 하는 것도 좋아보이네요!

handsone-u committed:
[#13] feat: 주문 수락/거절 기능

- adapter-in을 controller가 아닌 mq listener로 바꿈

handsone-u committed:
[#13] fix: 주문 수락/거절 기능

- 매장 도메인 주문 요청 로직을 작성하려고 다른 도메인들 보고 있는데, order 도메인에 주문 임시 저장 로직이 있더라고요.
- 그래서 임시 저장된 주문 조회-> 주문 엔티티 저장으로 바꿀게요.
- entity에 customerId 필드 추가

handsone-u commented:
주문 수락/거절의 리스너에서도 orderTransactionId를 받도록 하여 임시 주문을 이용하도록 변경했습니다.

handsone-u committed:
Revert "[#13] fix: 주문 수락/거절 기능"

This reverts commit 74b19c7e83ac7bc1aae3f613bc6da6750971a5f1.

handsone-u committed:
[#14] fix: order entity에 customerId 필드 추가

handsone-u committed:
[#14] add: 임시 주문 조회(어댑터 기능)

handsone-u commented:
위에 두 커밋 이슈를 잘못 연결했네요. #13 입니다.

JinTaekLim commented on src/main/java/com/baedal/order/adapter/out/persistence/manager/TempOrderManager.java:
```diff
@@ -17,9 +17,13 @@ private String getKey(String key) {
+
+  public SaveTempOrderDto getTempOrder(String orderTransactionId) {
+    Object object = orderRedisRepository.get(getKey(orderTransactionId));
+    return (SaveTempOrderDto) object;
```
Object 를 위와 같이 직접 캐스팅 했을 때 발생하는 예외 처리는 필요 없을까요?

JinTaekLim commented on src/main/java/com/baedal/order/application/service/OrderQueryService.java:
```diff
@@ -0,0 +1,19 @@
+
+@Service
+@RequiredArgsConstructor
+public class OrderQueryService {
```
어떤 역할을 담당하는 클래스인가요?

추가적으로, 현재 Service <-> Service 간 참조를 하고 있는데 이는 수정할 필요성이 있어보입니다!

handsone-u replied:
Service <-> Service 간 참조가 어느 클래스에 있나요?

handsone-u replied:
https://github.com/Project-Baedal/order/pull/12/commits/ed1776834a9d3a1af243580d4971d64e061491cd
주문 도메인 조회 서비스 용도입니다.

JinTaekLim replied:
잘못 확인했었나봅니다!

혹시 OrderService 에 구현하지 않고 OrderQueryService 를 별도로 구분한 이유가 있을까요?

handsone-u replied:
쿼리랑 커맨드를 서비스 계층에서 분리하려는 의도였습니다. orderService가 조금 커보여서 따로 만들었어요.

JinTaekLim replied:
쿼리는 RepositoryPort 에서 담당하는 만큼, 이를 Service 단에서 구분하는 것은 적합하지 않다는 생각이 먼저 드는 것같습니다.

때문에 분리하게 되더라도 Adapter 내부에서  처리하는 것이 맞지 않나 하는 의문입니다.

handsone-u committed:
[#14] add: 임시 주문 조회 기능 예외 처리
